### ISSUE TITLE
Dockerfile build fails at librocksdb-sys v5.11.3 because libclang missing

### ISSUE THREAD

**Issue by brenzi** [paritytech/substrate#1321]
expected:

```
docker build . 
```
succeeds

error:
```
   Compiling libp2p-tcp-transport v0.1.0 (https://github.com/libp2p/rust-libp2p?rev=d961e656a74d1bab5366d371a06f9e10d5f4a6c5#d961e656)
   Compiling libp2p-dns v0.1.0 (https://github.com/libp2p/rust-libp2p?rev=d961e656a74d1bab5366d371a06f9e10d5f4a6c5#d961e656)
   Compiling jsonrpc-http-server v9.0.0 (https://github.com/paritytech/jsonrpc.git#0d78b8f1)
   Compiling libp2p-kad v0.1.0 (https://github.com/libp2p/rust-libp2p?rev=d961e656a74d1bab5366d371a06f9e10d5f4a6c5#d961e656)
error: failed to run custom build command for `librocksdb-sys v5.11.3`
process didn't exit successfully: `/substrate/target/release/build/librocksdb-sys-ca5ab8b094d35b0e/build-script-build` (exit code: 101)
--- stdout
cargo:rerun-if-changed=build.rs
cargo:rerun-if-changed=rocksdb/
cargo:rerun-if-changed=snappy/

--- stderr
thread 'main' panicked at 'Unable to find libclang: "couldn\'t find any of [\'libclang.so\', \'libclang.so.*\', \'libclang-*.so\'], set the LIBCLANG_PATH environment variable to a path where one of these files can be found (skipped: [])"', libcore/result.rs:1009:5
note: Run with `RUST_BACKTRACE=1` for a backtrace.

warning: build failed, waiting for other jobs to finish...
error: build failed
The command '/bin/sh -c cargo build --$PROFILE' returned a non-zero code: 101
```

guess it's the same issue as #1240

**Comment by holygits** [comment]
The working Dockerfile is https://github.com/paritytech/substrate/blob/master/scripts/docker/Dockerfile

apline's cargo package uses stable but nightly is required to build substrate properly.
There is no rust nightly support to build on alpine, as a result the new Dockerfile is based on a debian distro which supports rust nightly.

I think you'll need to use the new Dockerfile

**Comment by chevdor** [comment]
Thanks for reporting @brenzi and sorry for the belated answer. I guess you may have noticed that: see TLDR

## TLDR:

The issue is solved for a while (fix in 527459c3).

## Details 

What do you mean `the working one` @holygits ? 😄 

Both Dockerfile work properly fine. They do however different things and I would say they are meant for different targets.

### CI Dockerfile

The Dockerfile you point out is rather new and it is the one used by the CI for a few months. 

Keypoint: This image does not build anything but packs an already built binary 'somewhere/somehow' in a docker image. This is a good solution for QA and CI as it is much faster to 'build' as the second option. This is however not such a great solution for users who want to build their own as they need to install and maintain full rust toolchain and as we just saw, dependencies that may come and go such as clang.

### Root Dockerfile
 
This one is the one the issue is related to. Initially (we are talking back a year ago), clang was not required. Back then, my goal was to offer any user (windows, linux, OSX, etc..) with *or without* any rust toolchain installed, the possibility to simply build and run their very own polkadot/substrate node.

The clang dependency became required afterward and was missing as @brenzi reported.
In the meantime, the missing clang has been added though.

I just checked and `clang` is now part of the 'root' `Dockerfile`:
```
$ cat Dockerfile | grep clang
	apt-get install -y cmake pkg-config libssl-dev git clang
```

I also tried building the image with `docker build .` and it did succeed as you can see below.

<details>
<summary>`docker build .` output</summary>

```
$ docker build . 
Sending build context to Docker daemon    161MB
Step 1/20 : FROM phusion/baseimage:0.10.2 as builder
 ---> 4ca439e72536
Step 2/20 : LABEL maintainer="chevdor@gmail.com"
 ---> Using cache
 ---> 96af844a383f
Step 3/20 : LABEL description="This is the build stage for Substrate. Here we create the binary."
 ---> Running in 6be9d8453b14
Removing intermediate container 6be9d8453b14
 ---> 5ab81a56e3be
Step 4/20 : ARG PROFILE=release
 ---> Running in ed5689abba57
Removing intermediate container ed5689abba57
 ---> 40552eb35ecf
Step 5/20 : WORKDIR /substrate
 ---> Running in a48330c346ee
Removing intermediate container a48330c346ee
 ---> cd10791447c4
Step 6/20 : COPY . /substrate
 ---> f6bb5f872d50
Step 7/20 : RUN apt-get update && 	apt-get upgrade -y && 	apt-get install -y cmake pkg-config libssl-dev git clang
 ---> Running in 1bdd9b8c50e7
<...a while later...>
Removing intermediate container a3d127ed5c54
 ---> 12afaf12bc73
Step 18/20 : EXPOSE 30333 9933 9944
 ---> Running in 57a61486966b
Removing intermediate container 57a61486966b
 ---> 9082556ec21b
Step 19/20 : VOLUME ["/data"]
 ---> Running in 8c177a5026bd
Removing intermediate container 8c177a5026bd
 ---> 6709532b9e65
Step 20/20 : CMD ["/usr/local/bin/substrate"]
 ---> Running in 4586cf00eb73
Removing intermediate container 4586cf00eb73
 ---> e4d54ad3eb0a
Successfully built e4d54ad3eb0a
```
</details>

I am closing this issue now @brenzi but feel free to re-opened or open a new issue if you run into any issue.
